glommio/src/executor/stall.rs

```diff
@@ -0,0 +1,534 @@
+use std::{
+    fmt,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    thread::{self, JoinHandle},
+    time::{Duration, Instant},
+};
+
+pub struct StallDetection<'a> {
+    executor: usize,
+    queue_handle: TaskQueueHandle,
+    queue_name: &'a str,
+    trace: backtrace::Backtrace,
+    budget: Duration,
+    overage: Duration,
+}
+
+impl fmt::Debug for StallDetection<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("StallDetection")
+            .field("executor", &self.executor)
+            .field("queue_handle", &self.queue_handle)
+            .field("queue_name", &self.queue_name)
+            .field("trace", &self.trace)
+            .field("budget", &self.budget)
+            .field("overage", &self.overage)
+            .finish()
+    }
+}
+
+impl fmt::Display for StallDetection<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "[stall-detector -- executor {}] task queue {} went over-budget: {:#?} (budget: \
+             {:#?}). Backtrace: {:#?}",
+            self.executor, self.queue_name, self.overage, self.budget, self.trace,
+        )
+    }
+}
+
+/// Trait describing what signal to use to trigger stall detection,
+/// how far past expected execution time to trigger a stall,
+/// and how to handle a stall detection once triggered.
+pub trait StallDetectionHandler: std::fmt::Debug + Send + Sync {
+    /// How far past the preemption timer should qualify as a stall
+    /// If None is returned, don't use the stall detector for this task queue.
+    fn high_water_mark(
+        &self,
+        queue_handle: TaskQueueHandle,
+        max_expected_runtime: Duration,
+    ) -> Option<Duration>;
+
+    /// What signal number to use; see values in libc::SIG*
+    fn signal(&self) -> u8;
+
+    /// Handler called when a task exceeds its budget
+    ///
+    /// **NOTE**: This callback is invoked from inside the reactor.
+    /// Do not perform blocking syscalls, acquire locks, spawn tasks
+    /// or otherwise block here.
+    fn stall(&self, detection: StallDetection<'_>);
+}
+
+/// Default settings for signal number, high water mark and stall handler.
+/// By default, the high water mark to consider a task queue stalled is set to
+/// 10% of the expected run time. The default handler will log a stack trace of
+/// the currently executing task queue. The default signal number is
+/// [`nix::libc::SIGUSR1`].
+#[derive(Debug)]
+pub struct LoggingStallDetectionHandler {
+    /// What signal to use to trigger stall detection
+    pub signal: u8,
+    /// What percentage of the original expected runtime to allow before
+    /// triggering a stall
+    pub overage_ratio: f64,
+    /// Absolute minimum duration before triggering a stall
+    pub min_duration: Duration,
+}
+
+impl LoggingStallDetectionHandler {
+    fn new(signal: u8, overage_ratio: f64, min_duration: Duration) -> Self {
```

Comment:
I think we should provide a concrete class with no settable properties, and allow extensions only through traits.

Reply:
I was pushing for a configurable class that provides `Default`